src/cis2/types.rs

```diff
@@ -0,0 +1,645 @@
+        let mut results = Vec::with_capacity(len.into());
+        for _ in 0..len {
+            results.push(MetadataUrl::deserial(source)?)
+        }
+        Ok(TokenMetadataQueryResponse::from(results))
+    }
+}
+
+/// A URL for the metadata.
+#[derive(Debug, Clone)]
+pub struct MetadataUrl {
+    /// The url encoded according to CIS2.
+    pub url:  String,
+    /// An optional checksum of the content found at the URL.
+    pub hash: Option<Hash>,
+}
+
+/// Deserialization for MetadataUrl according to the CIS2 specification.
+impl Deserial for MetadataUrl {
+    fn deserial<R: Read>(source: &mut R) -> Result<Self, ParseError> {
+        let len = source.read_u16()?;
+        let mut bytes = Vec::with_capacity(len.into());
+        for _ in 0..len {
+            bytes.push(source.read_u8()?)
+        }
+        let url = String::from_utf8(bytes)?;
+        let hash = Option::<[u8; 32]>::deserial(source)?.map(|b| b.into());
+        Ok(MetadataUrl { url, hash })
+    }
+}
+
+/// Smart contract logged event, part of the CIS2 specification.
+#[derive(Debug, Display)]
+pub enum Event {
+    /// Transfer of an amount of tokens
+    #[display(
+        fmt = "Transferred token with ID {} from {} to {}",
+        token_id,
+        "address_display(from)",
+        "address_display(to)"
+    )]
+    Transfer {
+        token_id: TokenIdVec,
+        amount:   TokenAmount,
+        from:     Address,
+        to:       Address,
+    },
+    /// Minting an amount of tokens
+    #[display(
+        fmt = "Minted token with ID {} for {}",
+        token_id,
+        "address_display(owner)"
+    )]
+    Mint {
+        token_id: TokenIdVec,
+        amount:   TokenAmount,
+        owner:    Address,
+    },
+    /// Burning an amount of tokens
+    #[display(
+        fmt = "Burned token with ID {} for {}",
+        token_id,
+        "address_display(owner)"
+    )]
+    Burn {
+        token_id: TokenIdVec,
+        amount:   TokenAmount,
+        owner:    Address,
+    },
+    /// Add/Remove an address as operator for some other address.
+    #[display(
+        fmt = "{} {} as operator for {}",
+        update,
+        "address_display(operator)",
+        "address_display(owner)"
+    )]
+    UpdateOperator {
+        update:   OperatorUpdate,
+        owner:    Address,
+        operator: Address,
+    },
+    /// Provide an URL with the metadata for a certain token.
+    #[display(
+        fmt = "Added metadata url {} ({}) for token with ID {}",
+        "metadata_url.url",
+        "display_hash(metadata_url.hash)",
+        token_id
+    )]
+    TokenMetadata {
+        token_id:     TokenIdVec,
+        metadata_url: MetadataUrl,
+    },
+    /// Custom event outside of the CIS2 specification.
+    #[display(fmt = "Unknown event: Event is not part of CIS2 specification")]
+    Unknown,
+}
+
+fn display_hash(hash_opt: Option<Hash>) -> String {
+    if let Some(hash) = hash_opt {
+        format!("with hash {}", hex::encode(hash))
```

Comment:
What `Hash` type is this? The one defined in the SDK (which it should be) has a Display implementation already and it is that one that should be used.